### PR TITLE
SQS Automated tests implementation

### PR DESCRIPTION
**Pull request by neckron**

### CODE REVIEW

File: src/main/java/com/avioconsulting/log4j/sqs/client/ConnectorClient.java [hunk 1]
```diff
@@ -78,29 +77,41 @@ private void initSQSClient(ConnectorClientAttributes attributes) {
...
-
-	public String getQueueURL(String queueName) {
-		if(this.queueUrl == null) {
-			this.queueUrl = this.awsSQSAsyncClient.getQueueUrl(queueName).getQueueUrl();
+	public void sendMessages(MessageRequestWrapper messageList, String processorType, String queueName, String largeMessageQueueName) {
+		if (ProcessorType.EXTENDED.name().equals(processorType)){
+			sendExtendedMessages(messageList,largeMessageQueueName);
+		} else if (ProcessorType.S3.name().equals(processorType)) {
+			initS3Client(connectorClientAttributes);
+			messageList.getPutObjectRequest().forEach(message -> this.awsS3Client.putObject(message));
```
Comment (adesjardin) [inline src/main/java/com/avioconsulting/log4j/sqs/client/ConnectorClient.java]: Any reason we want to put the init methods in the sendMessages codepath?  I know it will see the client is already initialized but why even go into those methods every time we try and send a message instead of initializing on creation of the client?

File: src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java [hunk 2]
```diff
@@ -21,55 +21,54 @@
...
-		UUID uuid = UUID.randomUUID();
-		logger.debug("Large Message UUID: {}", uuid);
-		String[] splitMessage = splitStringByByteLength(processorAttributes.getMessage(), "UTF-8", processorAttributes.getMaxMessageSize());
-		MessageRequestWrapper messageRequestWrapper = new MessageRequestWrapper();
+    @Override
+    public MessageRequestWrapper process(ProcessorAttributes processorAttributes) {
+        List<SendMessageRequest> sendMessageRequestList = new ArrayList<>();
+        logger.debug("Splitting large message");
+        UUID uuid = UUID.randomUUID();
+        logger.debug("Large Message UUID: {}", uuid);
```
Comment (adesjardin) [inline src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java]: Can we just combine these into a single debug log that provides the UUID of the large message?

File: src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java [hunk 3]
```diff
@@ -21,55 +21,54 @@
...
-			sendMessageRequest.setMessageBody(String.format("currentPart=%d|totalParts=%d|uuid=%s|message=%s", i + 1, splitMessage.length, uuid, splitMessage[i]));
-			sendMessageRequest.setMessageGroupId(uuid.toString());
-			sendMessageRequest.setQueueUrl(processorAttributes.getQueueUrl());
-			sendMessageRequestList.add(sendMessageRequest);
-			messageRequestWrapper.setSendMessageRequest(sendMessageRequestList);
-		}
-		return messageRequestWrapper;
-	}
+        for (int i = 0; i < splitMessage.length; i++) {
+            logger.debug("Sending message {} of {}", i + 1, splitMessage.length);
```
Comment (adesjardin) [inline src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java:35]: We should make sure to include the UUID here if we're going to log each part, otherwise anything more than a single request at a time is going to get hard to follow

File: src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java [hunk 4]
```diff
@@ -21,55 +21,54 @@
...
-	}
+    /**
+     * Split the input string into a list of strings, each with a maximum memory size of maxsize bytes.
+     *
+     * @param src      the source string
+     * @param encoding the encoding to use for the output string
+     * @param maxsize  the maximum size of the string parts in bytes
+     * @return an array of strings, each with a maximum length of maxsize bytes
+     */
+    public static String[] splitStringByByteLength(final String src, final String encoding, final int maxsize) {
```
Comment (adesjardin) [inline src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java]: Was there a reason for changing this to public?  Nevermind I see it was for testing purposes.  We probably should remove the static though, this is used from an instance method only, no reason for it to be static.

File: pom.xml [hunk 5]
```diff
@@ -19,6 +19,7 @@
         <aws.sdk.version>1.12.234</aws.sdk.version>
         <log4j.version>2.17.2</log4j.version>
         <junit.version>4.13.2</junit.version>
+        <commonsIO.version>2.11.0</commonsIO.version>
```
Comment (adesjardin) [inline pom.xml]: commons.io.verison to keep with our standard naming

File: Readme.md [hunk 6]
```diff
@@ -64,37 +70,62 @@ or
...
+==========================
+* Run automated test of this componen using the following command.
+
+```
+mvn clean test -DawsAccessKey=YOURAWSKEY /
+ -DawsSecretKey=YOURAWSSECRET / 
+ -DawsRegion=YOURAWSREGION -DawsBucketName=your-bucket /
+ -DawsQueueName=your-queue / 
+ -DawsLargeMessageQueueName=your-large-queue /
+ -DmaxMessageBytes=25600
```
Comment (adesjardin) [inline Readme.md]: Let's provide standard values for everything except accessKey and secretKey.   Also we should list what permissions the credential needs to have in AWS

Comment (adesjardin) [inline Readme.md]: Also 25600 needs to be 256 otherwise the test does not pass

**Review by adesjardin — CHANGES_REQUESTED** [review]

File: src/main/java/com/avioconsulting/log4j/sqs/client/ConnectorClient.java (hunk 1 above)
Comment (neckron) [reply src/main/java/com/avioconsulting/log4j/sqs/client/ConnectorClient.java]: Yes, let's check this scenario: when appender is used with `largeMessageMode="S3"` it's not required to get `awsSQS` nor `awsExtended` instantiated to append logs into s3. So, as there's already this logic implemented at method `sendMessages`:

```

if (ProcessorType.EXTENDED.name().equals(processorType)){
 .. send to sqs-extended
} else if (ProcessorType.S3.name().equals(processorType)) {
 .. send to s3
} else {
 ... send to sqs
}
``` 
I took the chance to not replicate if-else logic to decide which client should be created. 

Hope it's clear enough, I really try to explain it as better as I could :D

File: src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java (hunk 2 above)
Comment (neckron) [reply src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java]: Done

File: pom.xml (hunk 5 above)
Comment (neckron) [reply pom.xml]: Done

File: Readme.md (hunk 6 above)
Comment (neckron) [reply Readme.md]: Done

Comment (neckron) [reply Readme.md]: Done

File: src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java (hunk 3 above)
Comment (neckron) [reply src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java:35]: That is  just an informative log of the appender, not the actual message to be sent to sqs. If you are agree, I can add the uuid at here.

File: src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java (hunk 4 above)
Comment (neckron) [reply src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java]: Done

**Review by TimGonz-AVIO — APPROVED** [review]

File: src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java (hunk 3 above)
Comment (adammead) [reply src/main/java/com/avioconsulting/log4j/sqs/processor/FifoMessageProcessor.java:35]: I think it would still be good to log the UUID so you can tell in the appender logs that all of the log messages are for the same set of messages to SQS

File: src/main/java/com/avioconsulting/log4j/sqs/client/ConnectorClient.java (hunk 1 above)
Comment (neckron) [reply src/main/java/com/avioconsulting/log4j/sqs/client/ConnectorClient.java]: Resolved

**Review by adesjardin — APPROVED** [review]
